### DISCUSSION TITLE
Help with training a NER model that returns confidence score for each entity

### DISCUSSION THREAD

**Discussion by akdanwlf** [explosion/spaCy#9635] [Help: Coding & Implementations] [answered]
Hello Everyone,

I would appreciate it if anyone can help me with the following two things:
- [x]  Validate the config file for the task I am doing
- [x] Suggest how should I generate the labels JSON file to be used to initialize the spancat component.

**Task** 
I need to build a ner model that provides a confidence score with each entity it finds. 

**Approach** 
I am trying to train a spacy pipeline containing tok2vec, spancat, and ner. I have a custom labelled IOB format training and validation data which I converted to spacy binary format using  `python -m spacy convert ./train.iob ./ -n 10`.

I added the required configuration to include tok2vec, span-cat and ner, also updated span-cat to use tok2vec listener, like suggested here https://github.com/explosion/spaCy/discussions/9601#discussioncomment-1581177. But when I run the debug command `python -m spacy debug data config.cfg` to validate, I get the following error. 

> ValueError: [E143] Labels for component 'spancat' not initialized. This can be fixed by calling add_label, or by providing a representative batch of examples to the component's `initialize` method.

So I am trying to add the `initialize.component.spancat` and `initialize.component.spancat.labels` to the config file, but I couldn't find a source on how to generate the spancat.json file.

```
[initialize.components.spancat]

[initialize.components.spancat.labels]
@readers = "spacy.read_labels.v1"
path = "../data/spancat.json"
```

My config.cfg file looks like this:

```
[paths]
train = "../data/train.spacy"
dev = "../data/val-data.spacy"
vectors = null
init_tok2vec = null

[system]
gpu_allocator = null
seed = 0

[nlp]
lang = "en"
pipeline = ["tok2vec","spancat","ner"]
batch_size = 1000
disabled = []
before_creation = null
after_creation = null
after_pipeline_creation = null
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}

[components]

[components.spancat]
factory = "spancat"
max_positive = null
spans_key = "sc"
threshold = 0.5

[components.spancat.model]
@architectures = "spacy.SpanCategorizer.v1"

[components.spancat.model.tok2vec]
@architectures = "spacy.Tok2VecListener.v1"
width = ${components.tok2vec.model.encode.width}
upstream = "*"

[components.ner]
factory = "ner"
incorrect_spans_key = null
moves = null
update_with_oracle_cut_size = 100

[components.ner.model]
@architectures = "spacy.TransitionBasedParser.v2"
state_type = "ner"
extra_state_tokens = false
hidden_width = 64
maxout_pieces = 2
use_upper = true
nO = null

[components.ner.model.tok2vec]
@architectures = "spacy.Tok2VecListener.v1"
width = ${components.tok2vec.model.encode.width}
upstream = "*"

[components.tok2vec]
factory = "tok2vec"

[components.tok2vec.model]
@architectures = "spacy.Tok2Vec.v2"

[components.tok2vec.model.embed]
@architectures = "spacy.MultiHashEmbed.v2"
width = ${components.tok2vec.model.encode.width}
attrs = ["ORTH","SHAPE"]
rows = [5000,2500]
include_static_vectors = true

[components.tok2vec.model.encode]
@architectures = "spacy.MaxoutWindowEncoder.v2"
width = 96
depth = 4
window_size = 1
maxout_pieces = 3

[corpora]

[corpora.dev]
@readers = "spacy.Corpus.v1"
path = ${paths.dev}
max_length = 0
gold_preproc = false
limit = 0
augmenter = null

[corpora.train]
@readers = "spacy.Corpus.v1"
path = ${paths.train}
max_length = 0
gold_preproc = false
limit = 0
augmenter = null

[training]
dev_corpus = "corpora.dev"
train_corpus = "corpora.train"
seed = ${system.seed}
gpu_allocator = ${system.gpu_allocator}
dropout = 0.1
accumulate_gradient = 1
patience = 1600
max_epochs = 0
max_steps = 20000
eval_frequency = 200
frozen_components = []
annotating_components = []
before_to_disk = null

[training.batcher]
@batchers = "spacy.batch_by_words.v1"
discard_oversize = false
tolerance = 0.2
get_length = null

[training.batcher.size]
@schedules = "compounding.v1"
start = 100
stop = 1000
compound = 1.001
t = 0.0

[training.logger]
@loggers = "spacy.ConsoleLogger.v1"
progress_bar = false

[training.optimizer]
@optimizers = "Adam.v1"
beta1 = 0.9
beta2 = 0.999
L2_is_weight_decay = true
L2 = 0.01
grad_clip = 1.0
use_averages = false
eps = 0.00000001
learn_rate = 0.001

[training.score_weights]
ents_f = 1.0
ents_p = 0.0
ents_r = 0.0
ents_per_type = null

[pretraining]

[initialize]
vectors = "en_core_web_lg"
init_tok2vec = ${paths.init_tok2vec}
vocab_data = null
before_init = null
after_init = null

[initialize.lookups]
@misc = "spacy.LookupsDataLoader.v1"
lang = ${nlp.lang}
tables = ["lexeme_norm"]

[initialize.components]

[initialize.tokenizer]
```

Using Python v3.8 and spaCy v3.1.4

**Comment by polm** [comment, marked as answer]
Sorry you're having trouble with this, you might find it helpful to look at the [spancat example project](https://github.com/explosion/projects/tree/v3/experimental/ner_spancat) for reference.

> So I am trying to add the initialize.component.spancat and initialize.component.spancat.labels to the config file, but I couldn't find a source on how to generate the spancat.json file.

The file is just a JSON list of labels, like `["red", "green", "blue"]`. However normally you don't need the file unless you're using a streaming corpus or something. It could possibly be an issue if your annotations aren't being picked up for some reason.

Are you sure you need an NER component in addition to a spancat component? In the linked issue the user just asked if it was possible to include both in the same pipeline, which it is, but typically I would expect you to only need one or the other.

**Reply by akdanwlf** [reply]
@polm thanks for the reply.

I actually only need the NER component, but to get the confidence score along with each entity I found from the discussion https://github.com/explosion/spaCy/issues/5917 here I have to retrain my entire pipeline by adding spancat.

Can you suggest if there is any other way to get a confidence score for each NER result without using the spancategorizer?

**Reply by polm** [reply]
Spancat does not add confidence scores to an NER model, and it does not give you confidence scores for annotations from an NER component. What you can do is re-use NER annotations for the spancat, get NER-like output from spancat, and that output will include confidence scores.

If all you need is NER with confidence scores, you specifically should NOT use NER + spancat, just use a spancat instead of NER.

**Reply by akdanwlf** [reply]
Thanks for explaining this to me. I will try this and mark your recommendation as the answer, once I test it out.

**Reply by akdanwlf** [reply]
@polm I am afraid what you guessed seems to happen. 
> It could possibly be an issue if your annotations aren't being picked up for some reason.
I am using IOB format data and converting it to `.spacy` format using `python -m spacy convert ./train.iob ./ -n 10`

Does spancat component expect data in a different format?

**Reply by polm** [reply]
You can't just use `spacy convert` with spancat, you'll need to prep the training docs. The example project has a preprocessing script you can reference, but you can also do something like this:

```
import spacy
from spacy.tokens import DocBin
nlp = spacy.blank("en")

# first load the files made with spaCy convert
olddb = DocBin().from_disk("some/path")
# to hold converted docs
newdb = DocBin()

for doc in olddb.get_docs(nlp.vocab):
    doc.spans["sc"] = doc.ents
    newdb.add(doc)

newdb.to_disk("some/other/path.spacy")
```

**Reply by akdanwlf** [reply]
One more clarification before I train the model again. 

In the above code, we are setting the spans_key as `sc`. 

In the training config example from the project the spans_key is set to `grit` and values in training.score_weights are set to score on grit and null for sc.
```
.
.
.
spans_key = "grit"
.
.
.
[training.score_weights]
spans_grit_f = 1.0
spans_grit_p = 0.0
spans_grit_r = 0.0
spans_sc_f = null
spans_sc_p = null
spans_sc_r = null
.
.
.
```
Do I have to update spans_key to `sc` and update the training.score_weights with the values of `spans_grit_*` updated on `spans_sc_*` parameters?

**Reply by adrianeboyd** [reply]
The score weights is one part of the `spancat` config process that doesn't work that well out-of-the-box yet. With our templates and default configs, we don't have any way to automatically fill in the right default score weights if you change the key from `sc`.

The default span key with `init config` is `sc`. In the demo project we use a custom key an example throughout, but it's actually easier to use the `sc` default as long as you only have one `spancat` component.

If you're using `sc`, you can remove all the `grit` score weights and set `sc` ones to whatever values you'd like. `0.0` means it's shown in the training output but not factored into the overall score. `null` means it's not shown. You have to use `null` for the default `spans_sc_f` etc. in the `grit` demo because otherwise they show up as columns of useless `0.0` values.

**Reply by polm** [reply]
The choice of span key is arbitrary, `sc` is just the default. You do need the value in the config to match whatever you use in the training docs, including the scoring bits.

**Reply by akdanwlf** [reply]
Grateful for the help. I am able to complete my task. I would like to help anyone else with a similar problem, if you guys suggest I can contribute towards any documentation regarding this?

**Reply by polm** [reply]
Thanks for the offer! What kinds of changes to the docs did you have in mind / what would have helped you here?

**Reply by steve-solun** [reply]
@polm I think that adding a colab or a notebook with a simple example of labeled entities + spacy blank model + relevant config file showing a working example how to train a spancat model will be super helpful. I am investing so much time to gather the bits and bytes to make it work, without any luck. See my questions here and on stackoverflow, let's make a simple working example from A to Z.

**Reply by rmitsch** [reply]
@steve-solun Are you aware of the [`projects` tutorials](https://github.com/explosion/projects) on NER? If so and you still think these are insufficient, could you let me know how do you think we could improve those?

**Reply by steve-solun** [reply]
@rmitsch yes, I have checked all the docs and most of the projects are 1+ years old and don't contain a simple readable real world A-Z example of taking annotated data and training a spancat model.
I will be happy if you would share the tutorial in a colab and show people how to use it(I am not the first and not the last). It should be constantly updated and add more and more examples. It will populate the methods and make them much more efficient. You will get so much good valuable feedback.
